File: apps/python-api/app/store/clickhouse.py
import json
import struct
from datetime import datetime, timezone
from urllib.parse import quote

import httpx

from app.models import Sample, Tag
from app.store.net import new_client, raise_status, split_http_url, unix_ms

_ROW = struct.Struct("<qIfH")
_INSERT = (
    "INSERT INTO samples (ts, tag_id, value, quality) "
    "SELECT fromUnixTimestamp64Milli(ts), tag_id, value, quality "
    "FROM input('ts Int64, tag_id UInt32, value Float32, quality UInt16') FORMAT RowBinary"
)


def _ch_time(ts: datetime) -> str:
    utc = ts.astimezone(timezone.utc) if ts.tzinfo else ts.replace(tzinfo=timezone.utc)
    return utc.strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]


def _from_ms(ms: int) -> datetime:
    return datetime.fromtimestamp(ms / 1000, tz=timezone.utc)


def _ids(tag_ids: list[int]) -> str:
    return ",".join(str(int(i)) for i in tag_ids)


def _locf_sql(ids: str, at: str, bounded: bool) -> str:
    at_dt = f"toDateTime64('{at}', 3, 'UTC')"
    lower = f" AND s.ts >= {at_dt} - INTERVAL 3 HOUR" if bounded else ""
    return f"""
        SELECT toUnixTimestamp64Milli(max(s.ts)), s.tag_id, argMax(s.value, s.ts), argMax(s.quality, s.ts)
        FROM samples AS s
        WHERE s.tag_id IN ({ids}) AND s.ts <= {at_dt}{lower}
        GROUP BY s.tag_id
    """


def _parse_rows(blob: bytes, carried: bool) -> list[Sample]:
    if len(blob) % _ROW.size:
        raise RuntimeError("clickhouse rowbinary truncated")
    return [
        Sample(ts=_from_ms(ts), tag_id=tag_id, value=float(value), quality=quality, carried=carried)
        for ts, tag_id, value, quality in _ROW.iter_unpack(blob)
    ]


def _merge_range(tag_ids: list[int], head: list[Sample], tail: list[Sample]) -> list[Sample]:
    buckets: dict[int, list[Sample]] = {int(t): [] for t in tag_ids}
    extra: list[Sample] = []
    for sample in head:
        buckets.setdefault(sample.tag_id, []).append(sample)
    for sample in tail:
        if sample.tag_id in buckets:
            buckets[sample.tag_id].append(sample)
        else:
            extra.append(sample)
    out: list[Sample] = []
    for tag_id in tag_ids:
        out.extend(buckets[int(tag_id)])
    out.extend(extra)
    return out


class ClickHouseStore:
    name = "clickhouse"

    def __init__(self, url: str, database: str) -> None:
        origin, auth = split_http_url(url)
        self._db = database
        self._http = new_client(timeout=30.0, base_url=origin, auth=auth)
        self._insert_path = (
            f"/?database={quote(database)}&query={quote(_INSERT)}"
            "&async_insert=1&wait_for_async_insert=1&async_insert_busy_timeout_ms=200"
            "&async_insert_max_data_size=1048576"
        )

    async def ping(self) -> None:
        await self._query("SELECT 1")

    async def write(self, samples: list[Sample]) -> None:
        if not samples:
            return
        payload = bytearray(18 * len(samples))
        offset = 0
        for s in samples:
            _ROW.pack_into(payload, offset, unix_ms(s.ts), int(s.tag_id), float(s.value), int(s.quality))
            offset += 18
        blob = bytes(payload)
        last: Exception | None = None
        for attempt in range(2):
            try:
                resp = await self._http.post(
                    self._insert_path,
                    content=blob,
                    headers={"Content-Type": "application/octet-stream"},
                )
                raise_status(resp, "clickhouse write")
                return
            except httpx.TransportError as exc:
                last = exc
                if attempt:
                    break
        raise last or RuntimeError("clickhouse write failed")

    async def locf(self, tag_ids: list[int], at: datetime) -> list[Sample]:
        stamp = _ch_time(at)
        rows = _parse_rows(await self._query_binary(_locf_sql(_ids(tag_ids), stamp, True)), False)
        found = {s.tag_id for s in rows}
        missing = [int(i) for i in tag_ids if int(i) not in found]
        if missing:
            rows.extend(_parse_rows(await self._query_binary(_locf_sql(_ids(missing), stamp, False)), False))
        return rows

    async def range(self, tag_ids: list[int], start: datetime, end: datetime) -> list[Sample]:
        head = [s.as_carried() for s in await self.locf(tag_ids, start)]
        ids = _ids(tag_ids)
        left, right = _ch_time(start), _ch_time(end)
        tail = _parse_rows(
            await self._query_binary(
                f"""
                SELECT toUnixTimestamp64Milli(s.ts), s.tag_id, s.value, s.quality
                FROM samples AS s
                WHERE s.tag_id IN ({ids})
                  AND s.ts > toDateTime64('{left}', 3, 'UTC')
                  AND s.ts <= toDateTime64('{right}', 3, 'UTC')
                """
            ),
            False,
        )
        return _merge_range(tag_ids, head, tail)

    async def upsert_tags(self, tags: list[Tag]) -> None:
        if not tags:
            return
        body = "\n".join(json.dumps({"id": t.id, "name": t.name, "unit": t.unit}, separators=(",", ":")) for t in tags)
        resp = await self._http.post(
            "/",
            params={"database": self._db, "query": "INSERT INTO tags (id, name, unit) FORMAT JSONEachRow"},
            content=body,
        )
        raise_status(resp, "clickhouse tags")

    async def list_tags(self) -> list[Tag]:
        rows = await self._query("SELECT id, name, unit FROM tags ORDER BY id")
        return [Tag(id=int(r[0]), name=str(r[1]), unit=str(r[2])) for r in rows]

    async def close(self) -> None:
        await self._http.aclose()

    async def _query(self, sql: str) -> list:
        resp = await self._http.post(
            "/",
            params={"database": self._db, "query": sql + " FORMAT JSONCompact"},
        )
        raise_status(resp, "clickhouse query")
        return resp.json().get("data") or []

    async def _query_binary(self, sql: str) -> bytes:
        resp = await self._http.post(
            "/",
            params={"database": self._db, "query": sql + " FORMAT RowBinary"},
        )
        raise_status(resp, "clickhouse query")
        return resp.content

File: apps/go-api/internal/store/clickhouse.go
package store

import (
	"context"
	"time"

	"github.com/ClickHouse/clickhouse-go/v2"
	"github.com/ClickHouse/clickhouse-go/v2/lib/driver"

	"prism/go-api/internal/model"
)

type ClickHouse struct {
	conn driver.Conn
}

func NewClickHouse(dsn string) (*ClickHouse, error) {
	opts, err := clickhouse.ParseDSN(dsn)
	if err != nil {
		return nil, err
	}
	opts.MaxOpenConns = 16
	opts.MaxIdleConns = 16
	opts.DialTimeout = 5 * time.Second
	opts.Compression = &clickhouse.Compression{Method: clickhouse.CompressionLZ4}
	conn, err := clickhouse.Open(opts)
	if err != nil {
		return nil, err
	}
	return &ClickHouse{conn: conn}, nil
}

func (s *ClickHouse) Name() string { return "clickhouse" }

func (s *ClickHouse) Close() error { return s.conn.Close() }

func (s *ClickHouse) Ping(ctx context.Context) error { return s.conn.Ping(ctx) }

func (s *ClickHouse) Write(ctx context.Context, samples []model.Sample) error {
	if len(samples) == 0 {
		return nil
	}
	if err := s.insert(ctx, samples); err != nil {
		return s.insert(ctx, samples)
	}
	return nil
}

func (s *ClickHouse) insert(ctx context.Context, samples []model.Sample) error {
	// Short busy window coalesces the 8 write-ceiling workers into fewer parts
	// without parking each HTTP write for the default 200ms flush.
	ctx = clickhouse.Context(ctx, clickhouse.WithSettings(clickhouse.Settings{
		"async_insert":                 1,
		"wait_for_async_insert":        1,
		"async_insert_busy_timeout_ms": 10,
	}))
	batch, err := s.conn.PrepareBatch(ctx, "INSERT INTO samples (ts, tag_id, value, quality)")
	if err != nil {
		return err
	}
	for i := range samples {
		p := &samples[i]
		if err := batch.Append(p.TS.UTC(), p.TagID, float32(p.Value), p.Quality); err != nil {
			_ = batch.Abort()
			return err
		}
	}
	return batch.Send()
}

const chLocfLookback = 48 * time.Hour

const chLocfSQL = `
		SELECT s.ts, s.tag_id, s.value, s.quality
		FROM samples AS s
		WHERE (s.tag_id, s.ts) IN (
			SELECT t.tag_id, max(t.ts)
			FROM samples AS t
			WHERE t.tag_id IN ? AND t.ts <= ?
			GROUP BY t.tag_id
		)`

const chLocfBoundedSQL = `
		SELECT s.ts, s.tag_id, s.value, s.quality
		FROM samples AS s
		WHERE (s.tag_id, s.ts) IN (
			SELECT t.tag_id, max(t.ts)
			FROM samples AS t
			WHERE t.tag_id IN ? AND t.ts <= ? AND t.ts >= ?
			GROUP BY t.tag_id
		)`

func (s *ClickHouse) Locf(ctx context.Context, tagIDs []uint32, at time.Time) ([]model.Sample, error) {
	out, err := s.locf(ctx, tagIDs, at, true)
	if err != nil {
		return nil, err
	}
	missing := missingTagIDs(tagIDs, out)
	if len(missing) == 0 {
		return out, nil
	}
	rest, err := s.locf(ctx, missing, at, false)
	if err != nil {
		return nil, err
	}
	return append(out, rest...), nil
}

func (s *ClickHouse) locf(ctx context.Context, tagIDs []uint32, at time.Time, bounded bool) ([]model.Sample, error) {
	at = at.UTC()
	var (
		rows driver.Rows
		err  error
	)
	if bounded {
		rows, err = s.conn.Query(ctx, chLocfBoundedSQL, tagIDs, at, at.Add(-chLocfLookback))
	} else {
		rows, err = s.conn.Query(ctx, chLocfSQL, tagIDs, at)
	}
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanCH(rows, false)
}

func (s *ClickHouse) Range(ctx context.Context, tagIDs []uint32, from, to time.Time) ([]model.Sample, error) {
	head, err := s.Locf(ctx, tagIDs, from)
	if err != nil {
		return nil, err
	}
	rows, err := s.conn.Query(ctx, `
		SELECT s.ts, s.tag_id, s.value, s.quality
		FROM samples AS s
		WHERE s.tag_id IN ? AND s.ts > ? AND s.ts <= ?
		ORDER BY s.tag_id, s.ts`, tagIDs, from.UTC(), to.UTC())
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	tail, err := scanCH(rows, false)
	if err != nil {
		return nil, err
	}
	return mergeCHRange(tagIDs, head, tail), nil
}

func missingTagIDs(tagIDs []uint32, rows []model.Sample) []uint32 {
	found := make(map[uint32]struct{}, len(rows))
	for i := range rows {
		found[rows[i].TagID] = struct{}{}
	}
	var missing []uint32
	seen := make(map[uint32]struct{}, len(tagIDs))
	for _, id := range tagIDs {
		if _, dup := seen[id]; dup {
			continue
		}
		seen[id] = struct{}{}
		if _, ok := found[id]; !ok {
			missing = append(missing, id)
		}
	}
	return missing
}

func mergeCHRange(tagIDs []uint32, head, tail []model.Sample) []model.Sample {
	buckets := make(map[uint32][]model.Sample, len(tagIDs))
	for _, id := range tagIDs {
		buckets[id] = nil
	}
	for i := range head {
		head[i].Carried = true
		buckets[head[i].TagID] = append(buckets[head[i].TagID], head[i])
	}
	var extra []model.Sample
	for i := range tail {
		id := tail[i].TagID
		if _, ok := buckets[id]; ok {
			buckets[id] = append(buckets[id], tail[i])
		} else {
			extra = append(extra, tail[i])
		}
	}
	out := make([]model.Sample, 0, len(head)+len(tail))
	for _, id := range tagIDs {
		out = append(out, buckets[id]...)
	}
	return append(out, extra...)
}

func (s *ClickHouse) UpsertTags(ctx context.Context, tags []model.Tag) error {
	if len(tags) == 0 {
		return nil
	}
	batch, err := s.conn.PrepareBatch(ctx, "INSERT INTO tags (id, name, unit)")
	if err != nil {
		return err
	}
	for _, t := range tags {
		if err := batch.Append(t.ID, t.Name, t.Unit); err != nil {
			_ = batch.Abort()
			return err
		}
	}
	return batch.Send()
}

func (s *ClickHouse) ListTags(ctx context.Context) ([]model.Tag, error) {
	rows, err := s.conn.Query(ctx, `SELECT id, name, unit FROM tags ORDER BY id`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []model.Tag
	for rows.Next() {
		var t model.Tag
		if err := rows.Scan(&t.ID, &t.Name, &t.Unit); err != nil {
			return nil, err
		}
		out = append(out, t)
	}
	return out, rows.Err()
}

func scanCH(rows driver.Rows, carried bool) ([]model.Sample, error) {
	var out []model.Sample
	for rows.Next() {
		var s model.Sample
		var value float32
		if err := rows.Scan(&s.TS, &s.TagID, &value, &s.Quality); err != nil {
			return nil, err
		}
		s.Value = float64(value)
		s.Carried = carried
		out = append(out, s)
	}
	return out, rows.Err()
}
